### PR TITLE
Use thiserror for better error handling

### PR DESCRIPTION
**Pull request by SamrutGadde**
Apologies for the delay, but I've made the suggested changes as mentioned in a previous pull request (See #6376).

I've added just a few of the commits here to make it easier for review, let me know if you see any changes need to be made.

See #1910

### CODE REVIEW

**Comment by rbradford** [comment]
When you fix the build issue please make sure you squash them into the appropriate commits e.g. you can use `git autofixup` to do this followed by `git rebase -i main --autosquash`

**Comment by rbradford** [comment]
Please setup the git hooks as described in the CONTRIBUTING.md file.

**Comment by SamrutGadde** [comment]
For some reason the hook didn't catch that error, had to run the same command as in the test to check. It's fixed now.

File: devices/src/legacy/gpio_pl061.rs [hunk 1]
```diff
@@ -31,35 +32,24 @@ const GPIOAFSEL: u64 = 0x420; // Mode Control Select Register
                               // These registers can conceptually be treated as a 32-bit register, and PartNumber[11:0] is used to identify the peripheral.
                               // We are putting the expected values (look at 'Reset value' column from above mentioned document) in an array.
 const GPIO_ID: [u8; 8] = [0x61, 0x10, 0x14, 0x00, 0x0d, 0xf0, 0x05, 0xb1];
-// ID Margins
+// ID Marginsc
```
Comment (rbradford) [inline devices/src/legacy/gpio_pl061.rs:35]: You added a character here..

Comment (SamrutGadde) [reply devices/src/legacy/gpio_pl061.rs]: oops, fixed now.

File: api_client/src/lib.rs [hunk 2]
```diff
@@ -3,41 +3,27 @@
...
 use std::os::unix::io::RawFd;
+use thiserror::Error;
 use vmm_sys_util::sock_ctrl_msg::ScmSocket;
 
-#[derive(Debug)]
+#[derive(Debug, Error)]
 pub enum Error {
+    #[error("Error writing to or reading from HTTP socket: {0}")]
     Socket(std::io::Error),
+    #[error("Error writing to or reading from HTTP socket: {0}")]
```
Comment (rbradford) [inline api_client/src/lib.rs:15]: I know you copied from the `Display` implementation but this error is about sending file descriptors. Please fix as a separate commit.

**Comment by rbradford** [comment]
I think this is lgtm - @likebreath ?

**Comment by SamrutGadde** [comment]
apologies, added an accidental commit.

**Review by likebreath — COMMENTED** [review]
I had a quick pass and it looks good in general. I will need to take a closer look later this week. Let's run the CI workflows first.

**Review by likebreath — APPROVED** [review]
Thanks a lot for the contributions. LGMT.

**Comment by KshitijKapoor8** [comment]
For the sake of credit, @SamrutGadde and I worked on this task together.

**Pull request merged**